
## Results {-}

```{r}
knitr::read_chunk("R/results_query.R")
knitr::read_chunk("R/results_pvalues.R")
knitr::read_chunk("R/results_suppfiles.R")
```


```{r rna-seq-dynamics}

```

### HT-sequencing expression data in GEO database {-}

To assess the quality of RNAseq datasets we queried the NCBI GEO database for "expression profiling by high throughput sequencing" AND "human" OR "mouse".
NCBI GEO database shows exponentially growing number of high-throughput sequencing data set submissions between `r year(first_date)` and 2017 (Fig. \@ref(fig:queryfig)). 
First HT-seq data set that our query retrieved was submitted to GEO database at `r first_date` and our query was limited for datasets submitted until `r last_date`. 
This resulted in the retrieval of `r nrow(ds_redline)` datasets.

Approximately `r perc_mmhs` of GEO data sets (`r table(ds_redline$model)[1]`) contain data from mouse or human origin (Fig. \@ref(fig:queryfig), left facet).
The `r ppub[1]` (`r ppub_n_ci$ci[1]`, Poisson exact test) of human and mouse data sets are linked to their primary publication. 
Proportion of datasets with publications is lower when other taxa are used as model organisms `r ppub[2]` (`r ppub_n_ci$ci[2]`, Poisson exact test)(Fig. \@ref(fig:queryfig), right facet).
As expected, the gap between published GEO series and published articles is bigger for more recent submissions.

```{r missingsuppfiles}

```

For the human and mouse subset, we were able to identify supplementary files for `r perc_wsuppfile` (`r nrow(suppfilenames_present)`) of GEO series.  
The three percent of data sets lacking public supplementary files nevertheless show `r percent(ppub_n_ci_nopublicsupps$estimate, 0)` publication rate (`r ppub_n_ci_nopublicsupps$ci`) which is similar to data sets with supplementary files. 

(ref:queryfig) __Datasets submitted to NCBI GEO database featuring expression profiling by high-throughput sequencing__. GEO series submission dynamics and associated publications of HT-seq data. Left facet, submissions featuring human or mouse data, N = `r table(ds_redline$model)[1]`. Right facet, submissions of other taxa, N = `r table(ds_redline$model)[2]`. Dashed line, GEO series with publications.

```{r queryfig, fig.cap="(ref:queryfig)"}

```

### Compliance with GEO database submission requirements and data usability for reanalysis {-}

We sought to determine the compliance with the formal requirements for GEO data submission (https://www.ncbi.nlm.nih.gov/geo/info/seq.html). 
This pertains to the reproducibitly of data analysis. 
GEO archive has three required components: metadata, processed data, and raw data. 
Minimally these processed files are required to contain raw or normalized sequence read counts, or peak files of ChIP-Seq data. 
While, in the presence of raw data, supplementary data is not strictly neccessary for reproducibility, reproducing analysis from the raw data requires the reproduction of the full analysis, starting from re-aligning sequences to the genome, which can be practically unfeasible.

Supplemental HT-seq files uplodaded to NCBI GEO contain most commonly raw sequence data files with 'raw.tar' file extension and associated automatically generated 'filelist.txt' file (Fig. \@ref(fig:commonfilenames)).

(ref:commonfilenames) __Filename extensions and common stubs of SuppFileNames from HT-seq datasets submitted to NCBI GEO database__. (`r ref_labels("A", panel_label_case)`) Filename extensions of SuppFileNames. (`r ref_labels("B", panel_label_case)`) Common stubs of SuppFileNames. Stubs were obtained by removing GEO id string from lowercase converted SuppFileName. RAW.tar and its related filelist.txt file are automatically generated by GEO when compulsory raw sequence data is submitted as a supplemetary file. Stubs with N > 10 occurences are shown.

```{r commonfilenames, fig.cap="(ref:commonfilenames)", fig.asp=0.7}

```

Raw sequence read data is required part of GEO submission, when raw data is provided as supplemental raw.tar file, GEO submits raw data to SRA on behalf of user, this happens in `r supp_raw_perc` of submissions.

Other common files contain annotations 'readme.txt' and processed data: commonly sequence read counts, gene expression analysis results or ChIP-Seq peak counts. 

Sequence read counts can be either as raw read counts or normalised read counts.
Gene expression analysis of HT-seq data is commonly performed by fitting models to discrete sequence read count data.
Normalised sequence read counts are used for plotting and visual evaluation of the analysis results, but cannot be used to directly reproduce gene expression analysis. 

However, tables containing top-ranked genes/features (toptables) from gene expression analysis with complete set of raw P values allow direct evaluation of the data and analysis quality underlying tested hypothesis.

```{r out-strings}

```


(ref:filesofinterest) Distribution of supplementary files per GEO Accession before and after filtering supplementary files for download. Left, all supplementary files. Right, supplementary files to be downloaded after filtering for tabular formats. Numbers show total number of files, respectively, before and after filtering.

```{r filesofinterest, fig.cap='(ref:filesofinterest)', out.width='50%'}

```

GEO datasets contain most commonly one to two supplementary files per Accession and up to `r tail(suppfall_per_acc$Files, 1)` files per Accession (Fig. \@ref(fig:filesofinterest)).
To reduce number of files to be downloaded from GEO database, based on file extensions and filenames, we first excluded from the dataset files that most plausibly don't contain final processed data in a tabular format, such as binary files, annotations, raw data, graphics, and other non tabular files.
This filter reduced the number of files by `r files_oi_perc` (Fig. \@ref(fig:filesofinterest)) and GEO series by `r suppf_oi_perc`. 
Resulting dataset consisted of `r nrow(suppfiles_of_interest)` files from `r n_distinct(suppfiles_of_interest$Accession)` GEO series to be downloaded.

```{r loadst}

```


### Processed data

(ref:dims) __Number of samples and features of HT-seq experiments.__ (`r ref_labels("A", panel_label_case)`) Mean number of samples was `r round(n_samples$samples_mean)` and median `r n_samples$samples_median`. Most frequent sample size was `r n_samples$samples_Mode`. Mean number of features was `r paste(floor(n_samples$features_mean))` and median `r n_samples$features_median`. (`r ref_labels("B", panel_label_case)`) Number of features in GEO series where supplementary file with p values was uploaded (dark grey) compared to all GEO series with supplementary files (light grey). (`r ref_labels("C", panel_label_case)`) Number of samples in GEO series where p values were available (dark grey) compared to all series with supplementary files (light grey). Number of samples was taken from series matrix files and number of features is number of rows in supplementary files. Dashed line denotes cutoff for complete sets of features/p values.

```{r plotdims, fig.cap='(ref:dims)'}

```

